File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CsrSegmentMax.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Neighbourhood max-aggregation over a graph in CSR format: the GraphSAGE-max primitive.
 *
 * <p>For each target node {@code i} (row), collects the feature vectors of its source
 * neighbours {@code j} (given by {@code colIdx} entries in row {@code i}), then takes the
 * element-wise maximum:
 * <pre>
 *   out[i, f] = max_{k: rowPtr[i] <= k < rowPtr[i+1]} X[colIdx[k], f]
 * </pre>
 * This is the max-aggregator in GraphSAGE (Hamilton et al. 2017) and is used for
 * inductive representation learning on large graphs.  Rows with no neighbours
 * (empty adjacency row) produce {@code -Infinity} (or the backend's corresponding
 * sentinel) in the output.
 *
 * <p>C++ op name: {@code csr_segment_max}
 * <ul>
 *   <li>Inputs:  colIdx[nnz] (INT32), rowPtr[rows+1] (INT32), X[n, f] (floating)</li>
 *   <li>IArgs:   rows</li>
 *   <li>Output:  out[rows, f] — per-segment (per-row) feature max</li>
 * </ul>
 *
 * <p>{@code colIdx} and {@code rowPtr} are integer/structural and receive zero gradients.
 * Only {@code X} is differentiable; the backward pass is handled by {@link CsrSegmentMaxBp}.
 */
public class CsrSegmentMax extends DynamicCustomOp {

    /** Number of rows (segments / target nodes). Stored so doDiff can use it. */
    private long rows;

    /** No-arg constructor required for op-registry reflection. */
    public CsrSegmentMax() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param colIdx  1D [nnz] INT32 column (source-node) indices
     * @param rowPtr  1D [rows+1] INT32 row (segment) pointers
     * @param X       2D [n, f] dense node feature matrix (floating dtype)
     * @param rows    number of target-node rows (segments)
     */
    public CsrSegmentMax(INDArray colIdx, INDArray rowPtr, INDArray X, long rows) {
        super(new INDArray[]{colIdx, rowPtr, X}, null);
        this.rows = rows;
        addIArgument(rows);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd      the SameDiff graph
     * @param colIdx  SDVariable [nnz] INT32 column indices
     * @param rowPtr  SDVariable [rows+1] INT32 row pointers
     * @param X       SDVariable [n, f] dense node features
     * @param rows    number of rows (segments)
     */
    public CsrSegmentMax(SameDiff sd, SDVariable colIdx, SDVariable rowPtr,
                         SDVariable X, long rows) {
        super(sd, new SDVariable[]{colIdx, rowPtr, X});
        this.rows = rows;
        addIArgument(rows);
    }

    @Override
    public String opName() { return "csr_segment_max"; }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        // Output dtype matches X (input[2])
        return Collections.singletonList(dataTypes.get(2));
    }

    /**
     * Backward pass for csr_segment_max — delegates to {@link CsrSegmentMaxBp}.
     *
     * <p>The gradient is a masked scatter: {@code dX[j, f] = gradOut[i, f]} if node
     * {@code j} was the max-winning neighbour of row {@code i} at feature dimension
     * {@code f}, or zero otherwise.  Ties are broken by the native op (typically the
     * first occurrence wins).
     *
     * @param grads upstream gradients; {@code grads.get(0)} is d(loss)/d(out[rows, f])
     * @return gradients for [colIdx, rowPtr, X]
     */
    @Override
    public List<SDVariable> doDiff(List<SDVariable> grads) {
        SDVariable g = grads.get(0);

        long r = iArguments.size() > 0 ? iArguments.get(0) : this.rows;

        SDVariable colIdxArg = arg(0);
        SDVariable rowPtrArg = arg(1);
        SDVariable xArg      = arg(2);

        SDVariable dX = new CsrSegmentMaxBp(sameDiff, colIdxArg, rowPtrArg, xArg, g, r)
                .outputVariable();

        return Arrays.asList(
                sameDiff.zerosLike(colIdxArg),
                sameDiff.zerosLike(rowPtrArg),
                dX);
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/csr_segment_max.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// csr_segment_max — GraphSAGE-style max neighbor aggregation over CSR graph.
//
// For each row i and feature dimension feat:
//   out[i, feat] = max over k in [rowPtr[i], rowPtr[i+1]) of X[colIdx[k], feat]
// Empty row → out[i, :] = 0.
//
// Inputs:
//   [0] colIdx  [nnz]   int    — CSR column indices (neighbour node ids)
//   [1] rowPtr  [rows+1] int   — CSR row pointer array
//   [2] X       [n, f]  float  — dense node feature matrix
// IArgs:
//   [0] rows  — number of rows (output nodes)
// Outputs:
//   [0] out   [rows, f]  float  — aggregated features
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_csr_segment_max)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_segment_max.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(csr_segment_max, 3, 1, false, 0, 1) {
  auto colIdx = INPUT_VARIABLE(0);  // [nnz]    int
  auto rowPtr = INPUT_VARIABLE(1);  // [rows+1] int
  auto X      = INPUT_VARIABLE(2);  // [n, f]   float

  const sd::LongType rows = INT_ARG(0);

  REQUIRE_TRUE(colIdx->rankOf() == 1, 0,
               "csr_segment_max: colIdx must be rank-1, got %d", colIdx->rankOf());
  REQUIRE_TRUE(rowPtr->rankOf() == 1, 0,
               "csr_segment_max: rowPtr must be rank-1, got %d", rowPtr->rankOf());
  REQUIRE_TRUE(X->rankOf() == 2, 0,
               "csr_segment_max: X must be rank-2 [n, f], got %d", X->rankOf());
  REQUIRE_TRUE(rowPtr->lengthOf() == rows + 1, 0,
               "csr_segment_max: rowPtr length must be rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)rowPtr->lengthOf());

  auto out = OUTPUT_NULLIFIED(0);  // [rows, f]

  sd::ops::helpers::csr_segment_max(*colIdx, *rowPtr, *X, *out, rows);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(csr_segment_max) {
  auto X      = INPUT_VARIABLE(2);
  const sd::LongType rows = INT_ARG(0);
  const sd::LongType f    = X->sizeAt(1);

  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(
      X->dataType(), 'c', {rows, f});
  return SHAPELIST(outShape);
}

DECLARE_TYPES(csr_segment_max) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})    // colIdx
      ->setAllowedInputTypes(1, {ALL_INTS})    // rowPtr
      ->setAllowedInputTypes(2, {ALL_FLOATS})  // X
      ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_csr_segment_max)
